### DISCUSSION TITLE
UnicodeEncodeError: 'ascii' codec can't encode character '\xe4' in position 61: ordinal not in range(128)

### DISCUSSION THREAD

**Discussion by hartman** [pygments/pygments#2478] [General]
As reported at https://phabricator.wikimedia.org/T342117

One of the users reported an issue where he would receive the error 
```
*** Error while highlighting:
UnicodeEncodeError: 'ascii' codec can't encode character '\xe4' in position 61: ordinal not in range(128)
(file "/wiki/extensions/SyntaxHighlight_GeSHi/pygments/pygmentize/pygments/cmdline.py", line 189, in main_inner)
*** If this is a bug you want to report, please rerun with -v.
```

This was caused by running with the -V command. so here:
```
if argns.V:
        print('Pygments version %s, (c) 2006-2023 by Georg Brandl, Matthäus '
              'Chajdas and contributors.' % __version__)
        return 0
```

Removing the ä solves the problem. Likely triggered by the locale encoding used by the environment.

**Comment by antonyftp** [comment]
I'm the person who opened the ticket on MediaWiki phabricator. If you need any informations please let me know

**Comment by jeanas** [comment]
I can reproduce with

```
$ LC_ALL=C PYTHONUTF8=0 pygmentize -V

*** Error while highlighting:
UnicodeEncodeError: 'ascii' codec can't encode character '\xe4' in position 61: ordinal not in range(128)
   (file "/home/jean/repos/pygments/pygments/cmdline.py", line 188, in main_inner)
*** If this is a bug you want to report, please rerun with -v.
```

I'm surprised that `PYTHONUTF8=0` was needed but I suspect that my Fedora system is enabling UTF-8 mode by default sooner than PEP 686 prescribes (Python 3.15).

What are your locale settings? You can view them with the `locale` command.

Ultimately, all Pygments is doing “wrong“ is printing a non-ASCII character. I don't think we can count that as a bug. You just need an environment that is not configured in a legacy locale like the C locale where non-ASCII characters aren't supported. (Non-UTF-8 locales must die!)

**Comment by jeanas** [comment]
Hm, I just saw on the Mediawiki tracker that the locale settings are

```
LANGUAGE=en_US:en
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
```

`LC_ALL` is supposed to override everything, and is set to `en_US.UTF-8`, so that's strange. I also read that you are using a custom Docker image. Maybe the `en_US.UTF-8` locale isn't installed in that image? Does it appear in the output of `locale -a`?

**Comment by jeanas** [comment]
(Converted this to a discussion since it's not a problem in Pygments itself.)

**Reply by hartman** [reply]
Even if it is not a bug, an improvement might be possible. For instance if UTF-8 is a hard requirement, a simple:

```python
import locale

if locale.getencoding().upper() != 'UTF-8':
    logging.error('Pygments requires a UTF-8 environment');
```

Would be more understandable to end users I think.

**Reply by jeanas** [reply]
Well, you could say the same of each and every existing CLI written in Python in the world... Or even not written in Python...

(Note that we don't require a UTF-8 locale, we just need *some* encoding for the output (be it UTF-8, UTF-16, latin1, or whatever) that supports the characters we want to output.)
